### PR TITLE
DHS Feed v2

### PR DESCRIPTION
**Pull request by BEAdi**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4405, https://jira-hq.paloaltonetworks.local/browse/CIAC-4465

## Description
Enhancement to support TAXII v2 for DHS feed.
Currently we don't have an instance (requested one [here](https://jira-hq.paloaltonetworks.local/browse/U42ENGIS-71)), so it is not tested yet.

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            default_api_root=default_api_root,
+        )
+        client.initialise()
+
+        if command == 'test-module':
+            return_results(command_test_module(client, limit, fetch_full_feed))
+
+        elif command == 'fetch-indicators':
+            if fetch_full_feed:
+                limit = -1
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:191]: Per my previous comment, this should be removed.

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+
+            last_run_indicators = get_feed_last_run()
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:193]: To be honest, the `get_feed_last_run` function is not needed anymore, and can be deprecated - please use `demisto.getLastRun()`.

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            indicators, last_run_indicators = fetch_indicators_command(client, limit, last_run_indicators, initial_interval,
+                                                                       fetch_full_feed)
+            for iter_ in batch(indicators, batch_size=2000):
+                demisto.createIndicators(iter_)
+
+            set_feed_last_run(last_run_indicators)
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:199]: Same comment as above, this should be replaced with `demisto.setLastRun()`.

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            return_results(commands[command](client, **args))
+
+        else:
+            raise NotImplementedError(f'{command} command is not implemented.')
+
+    except Exception as error:
+        if isinstance(error, requests.exceptions.SSLError):
+            return_error('Encountered an HTTPS certificate error. This error can be ignored by enabling '
+                         '"Trust any certificate (not secure)" in the instance configuration.', error)
+        if isinstance(error, requests.HTTPError):
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:211]:
```suggestion
        elif isinstance(error, requests.HTTPError):
```

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    return last_fetch_time or initial_interval
+
+
+def get_indicators_command(client: Taxii2FeedClient, raw: str = 'false', limit: str = '10', added_after: str = '20 days'):
+    """
+    Fetch indicators from TAXII 2 server
+    :param client: Taxii2FeedClient
+    :param raw: When set to 'true' will return only rawJSON
+    :param limit: upper limit of indicators to fetch
+    :param (Optional) added_after: added after time string in parse_date_range format
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:94]: It's not really optional
```suggestion
    :param added_after: added after time string in parse_date_range format
```

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+import datetime
+
+''' CONSTANTS '''
+
+COMPLEX_OBSERVATION_MODE_SKIP = 'Skip indicators with more than a single observation'
+
+''' HELPER FUNCTIONS '''
+
+
+def assert_incremental_feed_params(fetch_full_feed: bool, is_incremental_feed: bool):
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:11]: This can be removed.

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def command_test_module(client: Taxii2FeedClient, limit: int, fetch_full_feed: bool):
+    if client.collections:
+        if fetch_full_feed:
+            if limit and limit != -1:
+                return 'Configuration Error - Max Indicators Per Fetch is disabled when Full Feed Fetch is enabled'
```
Comment (DeanArbel) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py:24]: This can be removed.

File: Packs/FeedDHS/ReleaseNotes/2_0_0.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+
+#### Integrations
+##### DHS Feed v2
+- Added support for TAXII V2 for the DHS feed.
```
Comment (DeanArbel) [inline Packs/FeedDHS/ReleaseNotes/2_0_0.md:4]: It's a new integration, so please add proper release notes.

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Looks good overall. See my comments below.

File: Packs/ApiModules/Scripts/TAXII2ApiModule/TAXII2ApiModule.py [hunk 9]
```diff
@@ -1025,13 +1031,27 @@ def poll_collection(
...
                 envelope = v20.as_pages(get_objects, per_request=page_size, **kwargs)
             else:
-                envelope = get_objects(limit=page_size, **kwargs)
+                demisto.debug('In v2.1, calling v21_get_objects')
+                envelope = self.v21_get_objects(limit=page_size, **kwargs)
             if envelope:
                 types_envelopes[obj_type] = envelope
         return types_envelopes
 
+    def v21_get_objects(self, accept="application/taxii+json;version=2.1", **filter_kwargs):
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/ApiModules/Scripts/TAXII2ApiModule/TAXII2ApiModule.py:1042]: Done

File: Packs/FeedDHS/ReleaseNotes/2_0_0.md (hunk 8 above)
Comment (BEAdi) [reply Packs/FeedDHS/ReleaseNotes/2_0_0.md]: Updated

**Comment by ShirleyDenkberg** [comment]
@DeanArbel Doc review completed.

**Review by Ni-Knight — APPROVED** [review]

**Review by yuvalbenshalom — APPROVED** [review]
For the TAXII2ApiModule changes.

File: Packs/FeedDHS/Integrations/DHSFeedV2/README.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+In the *dhs-get-indicators* command:
+
+* *DHS.type* - this output was replaced by *DHS.Indicators.type*.
+* *DHS.value* - this output was replaced by *DHS.Indicators.value*.
+* *DHS.tlp* - this output was removed.
+
+## Additional Considerations for this version
+
+Use this version if your certificate supports TAXII 2 protocol.
```
Comment (yuvalbenshalom) [inline Packs/FeedDHS/Integrations/DHSFeedV2/README.md:253]:
```suggestion
Use this version if your certificate supports TAXII 2 protocol.
## Known Limitations

First Fetch Time Parameter can be configured for a maximum of 48 hours, due to limitations in DHS TAXII2 API. Therefore, It is not possible to fetch indicators that has been last seen in the feed more than 48 hours in the past.

```

**Review by yuvalbenshalom — CHANGES_REQUESTED** [review]
Good job.
See my comments.

File: Packs/FeedDHS/Integrations/DHSFeedV2/README.md (hunk 10 above)
Comment (BEAdi) [reply Packs/FeedDHS/Integrations/DHSFeedV2/README.md:255]: Added

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+    - defaultValue: '10'
+      description: Maximum number of indicators to return. Default is 10. If you are increasing this value, make sure 'execution-timeout' argument is also increased.
+      name: limit
+    - defaultValue: '24 hours'
+      description: Fetch only indicators that were added to the server after the given
+        time. Provide a <number> and <time unit> of type minute/hour/day. For
+        example, 1 minute, 12 hour, 24 days. Limited to 48 hours.
+      name: added_after
+    description: Allows you to test your feed and to make sure you can fetch indicators
+      successfully. Due to API limitations this command may take a long time to run, make sure 'execution-timeout' argument is increased.
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml]: Done

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+      defaultValue: 'false'
+      description: Will return only the rawJSON of the indicator object.
+      name: raw
+      predefined:
+      - 'true'
+      - 'false'
+    - defaultValue: '10'
+      description: Maximum number of indicators to return. Default is 10. If you are increasing this value, make sure 'execution-timeout' argument is also increased.
+      name: limit
+    - defaultValue: '24 hours'
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml:210]: Done

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+  required: false
+  type: 1
+- defaultvalue: '240'
+  display: Feed Fetch Interval
+  name: feedFetchInterval
+  required: false
+  type: 19
+- additionalinfo: 'The time interval for the first fetch (retroactive) in the following format: <number> <time
+    unit> of type minute/hour/day. For example, 1 minute, 12 hour. Limited to 48 hours.'
+  defaultvalue: 24 hours
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml:97]: Done

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+            if limit <= 0:
+                break
+
+    return indicators, last_run_ctx
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def command_test_module(client: Taxii2FeedClient):
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.py]: Changed

File: Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml [hunk 15]
```diff
@@ -93,43 +93,21 @@ configuration:
   required: false
   type: 19
 - additionalinfo: 'The time interval for the first fetch (retroactive) in the following format: <number> <time
-    unit> of type minute/hour/day/year. For example, 1 minute, 12 hour.'
+    unit> of type minute/hour/day. For example, 1 minute, 12 hour. Limited to 48 hours.'
```
Comment (ShirleyDenkberg) [inline Packs/FeedDHS/Integrations/DHSFeedV2/DHSFeedV2.yml:96]: If it is more than 1 (hour, minute, day) will it be hours/minutes/days? For example 3 minutes. 12 hours.

**Comment by ShirleyDenkberg** [comment]
@yuvalbenshalom @DeanArbel @Ni-Knight Doc review completed.

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**